### ISSUE TITLE
What is ERTS_SYM_DIR?

### ISSUE THREAD

**Issue by jordanhubbard** [elixir-nx/nx#294]
Not suggesting the following diff, but it's the only thing that moves the compile past the fact that the erts subdirectory is not checked or called out as an explicit dependency in README.md (not sure why erlang's erts directory is being referenced at all).

```mix run``` works, on the other hand.
 
```
@@ -37,8 +37,8 @@ all: symlinks
 symlinks: $(TENSORFLOW_DIR)
        @ rm -f $(TENSORFLOW_EXLA_DIR)
        @ ln -s "$(MIX_CURRENT_PATH)/$(EXLA_DIR)" $(TENSORFLOW_EXLA_DIR)
-       @ rm -f $(ERTS_SYM_DIR)
-       @ ln -s "$(ERTS_INCLUDE_DIR)" $(ERTS_SYM_DIR)
+#      @ rm -f $(ERTS_SYM_DIR)
+#      @ ln -s "$(ERTS_INCLUDE_DIR)" $(ERTS_SYM_DIR)

```

**Comment by seanmor5** [comment]
To compile the NIF we need to link ERTS includes into the TF directory because Bazel doesn't work very well with global dependencies. 

The directory is included (usually) in standard Erlang installations. Sometimes you'll need to install `erlang-dev` as a separate dependency. Try installing `erlang-dev` and everything should work out!

**Comment by jordanhubbard** [comment]
Sorry, I think I somehow failed to make the more important fundamental points I was trying to make:

1.  I'm compiling all of this from source, including OTP, so there is no package dependency chain here.  The standard erlang install target, up to and including some of the more battery-included options like "install-docs", does not put an erts directory anywhere in /usr/local.
2. Even if it (or the erlang-dev package) did, this is not called out as a dependency for exla anywhere in the documentation so it's a "failed dependency" from that standpoint, at least.
3. Even with all of the above aside, the ```symlinks``` target in the Makefile never has the opportunity to succeed because ERTS_INCLUDE_DIR is not set or documented as needing to be set.   Ironically, relying on ERL_TOP, which *is* documented, would have a much greater chance of working, though this is now dragging in erlang to be even more explicit.

TL;DR:  If the erlang runtime is needed explicitly, why not just check it out in the same fashion as tensorflow (hardly a tiny dep) is being checked out?

**Comment by seanmor5** [comment]
`ERTS_INCLUDE_DIR` is automatically set by the Elixir make compiler to the Erlang/OTP install location so it doesn't usually need to be explicitly set. Although I'm not sure if that would be the case in this situation.

I'll document the need for `erlang-dev` as well. I don't have experience building Erlang/OTP from source, so I'm not sure what differences there would be w.r.t the EXLA build process. I can try and take some time tomorrow to get it up and running and document the process.

**Comment by jordanhubbard** [comment]
OK, let's also perhaps come up with some common notation which takes into account the difference between the build system "drivers" here.  We have, as you say, mix(1) as one build driver which is, in turn, driven from the top-level project make(1) driver which exposes just the `setup` and `test` targets.  I'll refer to those two as `Make(T,setup) -> Mix(S, deps.get)` and `Make(T,test)-> Mix(S, text)` where `T=$TOPDIR` and `S={nx, exla}`.  Make sense?  Those work fine.

The one we're getting hung up on is `Make(exla, all)`.  Should that even work?  It's not clear to me.  If that's considered "build target API" then it also suggests that `Make(T, all)` should also work, passing the appropriate make verb down to exla and skipping nx, since nx can only setup and test.

**Comment by elbow-jason** [comment]
@jordanhubbard what do you get when you run `erl` then `code:root_dir().`?

What I see:
```
$ erl
Erlang/OTP 23 [erts-11.1.3] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]

Eshell V11.1.3  (abort with ^G)
1> code:root_dir().               
"/home/elbow-jason/.asdf/installs/erlang/23.1.5"

```

**Comment by elbow-jason** [comment]
I'd also like to see `code:lib_dir('erts').`

**Comment by seanmor5** [comment]
I think I see what you're saying. Running `make` from the EXLA directory will not work. We rely on the `elixir_make` compiler to set all of the environment variables and invoke make for us, so attempting to build EXLA without running inside the mix project will always fail because the environment is not set up.

**Comment by jordanhubbard** [comment]
```
jkh@megamind-> erl
Erlang/OTP 24 [RELEASE CANDIDATE 1] [erts-11.1.8] [source-cd074622c9] [64-bit] [smp:32:32] [ds:32:32:10] [async-threads:1] [jit]

Eshell V11.1.8  (abort with ^G)
1> code:root_dir().               
"/usr/local/lib/erlang"
2> code:lib_dir('erts').
"/usr/local/lib/erlang/lib/erts-11.1.8"

```

Which is pretty much what I'd expect. to see, since the `install` target of otp puts everything in /usr/local and that's where all of the installations go.

To anticipate the next logical question:
```
jkh@megamind-> ls /usr/local/lib/erlang/
bin	   erts-10.7.1	  erts-11.1    Install	misc	     releases
COPYRIGHT  erts-10.7.2.1  erts-11.1.1  lib	PR.template  usr
doc	   erts-10.7.2.3  erts-11.1.8  man	README.md

```
Which also makes sense - that's where the various erts versions land when you build from source over time, since of course the version is mutable.

By the way, `megamind` is the name of the PC because it's a pretty high-spec PC+GPU and that seemed appropriate.  Had I named it after myself, `absentmind` would have been more accurate.

Anyway, you made me curious about the build-from-source vs curated pkg experience, of course, so I installed a fresh Ubuntu 20.04 VM and installed erlang-dev on it.  This is what I saw by contrast:

```
jkh@plex-> erl
Erlang/OTP 23 [erts-11.0.3] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1]

Eshell V11.0.3  (abort with ^G)
1> code:root_dir().               
"/usr/lib/erlang"
2>               
2> code:lib_dir('erts').
"/usr/lib/erlang/lib/erts-11.0.3"
```
Again, about what you'd expect from apt and the Linux convention that /usr is for both system and user packages.

**Comment by jordanhubbard** [comment]
I thought it might be easier to continue this discussion in a PR, since I don't want to seem like I'm just whining over here. 😃

**Comment by jordanhubbard** [comment]
This issue itself can be closed - we can take this up in the PR or see what the future build harness(es) look like in a split repo.  Thanks for the time and attention to what started as a hopefully small question. :)
